### PR TITLE
[JAVA][Spring]Optional params with delegate

### LINKED ISSUE #7179: [JavaSpring] Spring Boot codegen generates invalid code if optional date-time parameters used with delegate pattern
##### Description

Generating a server-stub with swagger codegen for a spring boot API. I'm looking at using the delgate pattern.

If I have a query parameter which is format date-time and required false, then the generated code has an error.  The API interface uses `Optional<OffsetDateTime>` as the variable type but the ApiDelegate Interface uses just `OffsetDateTime` so the (generated) call from one to the other fails

##### Swagger-codegen version

2.3.0-SNAPSHOT (I'm generating via the maven plugin, so I installed that version in my local repo)

##### Swagger declaration file content or url

```yaml
swagger: 2.0
info: 
    description: Test API
    version: 1.0
    title: Test
host: api.example.com
basePath: /test/
tags: 
- name: test
schemes: 
- https
consumes: 
- application/json
- application/xml
produces: 
- application/json
- application/xml
paths:
  /foo/:
    get:
      summary: example endpoint
      parameters:
        - $ref: '#/parameters/changedAfter'
      responses:
        '200':
            description: Success
parameters:
  changedAfter:
    name: changedAfter
    in: query
    description: Absolute or relative date-time string.
    required: false
    type: string
    format: date-time
```

##### Command line used for generation

`mvn -f pom_generate.xml clean compile`

using a pom_generate.xml of
```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>

	<groupId>org.springframework</groupId>
	<artifactId>gs-rest-service</artifactId>
	<version>0.1.0</version>
  
	<parent>
		<groupId>org.springframework.boot</groupId>
		<artifactId>spring-boot-starter-parent</artifactId>
		<version>1.5.9.RELEASE</version>
	</parent>
	<properties>
		<java.version>1.8</java.version>
		<maven.compiler.source>${java.version}</maven.compiler.source>
		<maven.compiler.target>${java.version}</maven.compiler.target>
		<springfox-version>2.7.0</springfox-version>
		<swagger.codegen.version>2.3.0-SNAPSHOT</swagger.codegen.version> 

		<jetty-version>9.2.15.v20160210</jetty-version>
		<slf4j-version>1.7.21</slf4j-version>
		<junit-version>4.12</junit-version>
		<servlet-api-version>2.5</servlet-api-version>
		<springfox-version>2.7.0</springfox-version>
		<jackson-version>2.8.9</jackson-version>
		<jackson-threetenbp-version>2.6.4</jackson-threetenbp-version>
		<spring-version>4.3.9.RELEASE</spring-version>
	</properties>
	
	<build>
		<sourceDirectory>src/main/java</sourceDirectory>
		<plugins>
			<plugin>
				<groupId>org.springframework.boot</groupId>
				<artifactId>spring-boot-maven-plugin</artifactId>
				<executions>
					<execution>
						<goals>
							<goal>repackage</goal>
						</goals>
					</execution>
				</executions>
			</plugin>
			<!-- Needed to create swagger bits in asynch manner -->
			<plugin>
				<groupId>io.swagger</groupId>
				<artifactId>swagger-codegen-maven-plugin</artifactId>
				<version>${swagger.codegen.version}</version>
				<executions>
					<execution>
						<id>common</id>
						<goals>
							<goal>generate</goal>
						</goals>
						<configuration>
							<inputSpec>src/spec/param.yaml</inputSpec>
							<modelPackage>com.example.api.models</modelPackage>
							<apiPackage>com.example.api</apiPackage>
							<output>${project.basedir}</output>					
							<language>spring</language>
							<invokerPackage>com.example.api</invokerPackage>
							<basePackage>com.example.api</basePackage>
							<withXml>true</withXml>
							<configOptions>
								<artifactId>foo</artifactId>
								<artifactDescription>Test API</artifactDescription>
								<title>Test API</title>
								<artifactUrl>https://api.example.com</artifactUrl>
								<groupId>com.example.api</groupId>
								<artifactVersion>1</artifactVersion>								
								<sourceFolder>src/gen/java/main</sourceFolder>
								<serializableModel>true</serializableModel>
								<dateLibrary>java8</dateLibrary>
								<java8>true</java8>
								<async>true</async>
								<library>spring-boot</library>		
								<delegatePattern>true</delegatePattern>
								<useBeanValidation>true</useBeanValidation>
								<useOptional>true</useOptional>	
								<hideGenerationTimestamp>true</hideGenerationTimestamp>								
							</configOptions>							
						</configuration>
					</execution>
				</executions>
			</plugin>
		</plugins>	
					
	</build>
	
	<dependencies>
		<dependency>
			<groupId>io.springfox</groupId>
			<artifactId>springfox-swagger2</artifactId>
			<version>${springfox-version}</version>
			<exclusions>
				<exclusion>
					<groupId>com.fasterxml.jackson.core</groupId>
					<artifactId>jackson-annotations</artifactId>
				</exclusion>
			</exclusions>
		</dependency>

		<dependency>
			<groupId>io.springfox</groupId>
			<artifactId>springfox-swagger-ui</artifactId>
			<version>${springfox-version}</version>
		</dependency>

		<dependency>
			<groupId>com.fasterxml.jackson.datatype</groupId>
			<artifactId>jackson-datatype-jsr310</artifactId>
			<version>${jackson-version}</version>
		</dependency>

		<dependency>
			<groupId>junit</groupId>
			<artifactId>junit</artifactId>
			<version>${junit-version}</version>
			<scope>test</scope>
		</dependency>
		<!-- <dependency> <groupId>javax.servlet</groupId> <artifactId>servlet-api</artifactId> 
			<version>${servlet-api-version}</version> </dependency> -->
		<dependency>
			<groupId>javax.validation</groupId>
			<artifactId>validation-api</artifactId>
			<version>1.1.0.Final</version>
			<scope>provided</scope>
		</dependency>

		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-web</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-tomcat</artifactId>
			<scope>provided</scope>
		</dependency>

		<dependency>
			<groupId>com.github.joschi.jackson</groupId>
			<artifactId>jackson-datatype-threetenbp</artifactId>
			<version>2.6.4</version>
		</dependency>

		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-test</artifactId>
			<scope>test</scope>
		</dependency>

		<dependency>
			<groupId>com.jayway.jsonpath</groupId>
			<artifactId>json-path</artifactId>
			<scope>test</scope>
		</dependency>
		<dependency>
			<groupId>joda-time</groupId>
			<artifactId>joda-time</artifactId>
			<version>2.9.9</version>
		</dependency>
	</dependencies>
	
	<repositories>
		<repository>
			<id>spring-releases</id>
			<url>https://repo.spring.io/libs-release</url>
		</repository>
	</repositories>
	<pluginRepositories>
		<pluginRepository>
			<id>spring-releases</id>
			<url>https://repo.spring.io/libs-release</url>
		</pluginRepository>
	</pluginRepositories>
</project>
```

##### Steps to reproduce

Generate the code and refresh eclipse. FooAPI.java includes:
```java
@Api(value = "foo", description = "the foo API")
public interface FooApi {

    FooApiDelegate getDelegate();

    @ApiOperation(value = "example endpoint", nickname = "fooGet", notes = "", tags={  })
    @ApiResponses(value = { 
        @ApiResponse(code = 200, message = "Success") })
    @RequestMapping(value = "/foo/",
        produces = { "application/json", "application/xml" }, 
        consumes = { "application/json", "application/xml" },
        method = RequestMethod.GET)
    default CompletableFuture<ResponseEntity<Void>> fooGet(@ApiParam(value = "Absolute or relative date-time string.") @Valid @RequestParam(value = "changedAfter", required = false) Optional<OffsetDateTime> changedAfter) {
        return getDelegate().fooGet(changedAfter);
    }

}
```

with the following error:

> The method fooGet(OffsetDateTime) in the type FooApiDelegate is not applicable for the arguments (Optional<OffsetDateTime>)

because FooApiDelegate.java defines as follows:
```java
public interface FooApiDelegate {

    Logger log = LoggerFactory.getLogger(FooApi.class);

    default Optional<ObjectMapper> getObjectMapper() {
        return Optional.empty();
    }

    default Optional<HttpServletRequest> getRequest() {
        return Optional.empty();
    }

    default Optional<String> getAcceptHeader() {
        return getRequest().map(r -> r.getHeader("Accept"));
    }

    /**
     * @see FooApi#fooGet
     */
    default CompletableFuture<ResponseEntity<Void>> fooGet(OffsetDateTime changedAfter) {
        if(getObjectMapper().isPresent() && getAcceptHeader().isPresent()) {
        } else {
            log.warn("ObjectMapper or HttpServletRequest not configured in default FooApi interface so no example is generated");
        }
        return CompletableFuture.completedFuture(new ResponseEntity<>(HttpStatus.NOT_IMPLEMENTED));
    }

}
```
with `OffsetDateTime changedAfter` instead of `Optional<OffsetDateTime> changedAfter`

##### Suggest a fix/enhancement

In the codegen, the javaSpring templates I see that api.mustache uses:

`{{#allParams}}{{>queryParams}}{{>pathParams}}{{>headerParams}}{{>bodyParams}}{{>formParams}}{{#hasMore}},{{/hasMore}}{{/allParams}}`

with queryParams.mustache being:
`{{#isQueryParam}}{{#useBeanValidation}}{{>beanValidationQueryParams}}{{/useBeanValidation}}@ApiParam(value = "{{{description}}}"{{#required}}, required = true{{/required}}{{#allowableValues}}, allowableValues = "{{#values}}{{{.}}}{{^-last}}, {{/-last}}{{#-last}}{{/-last}}{{/values}}"{{/allowableValues}}{{#defaultValue}}, defaultValue = "{{{defaultValue}}}"{{/defaultValue}}) {{#useBeanValidation}}@Valid{{/useBeanValidation}} @RequestParam(value = "{{baseName}}"{{#required}}, required = true{{/required}}{{^required}}, required = false{{/required}}{{#defaultValue}}, defaultValue="{{{defaultValue}}}"{{/defaultValue}}) {{>optionalDataType}} {{paramName}}{{/isQueryParam}}`

for the arguments while apiDelegate.mustache uses just:
`{{#allParams}}{{^isFile}}{{{dataType}}}{{/isFile}}{{#isFile}}MultipartFile{{/isFile}} {{paramName}}{{#hasMore}},{{/hasMore}}{{/allParams}}`

So I guess it might be as simple as replacing the `dataType` in apiDelegate.mustache with `>optionalDataType`  Seems to fix it for me, so I'll create a PR

What is also confusing is that it seems to use the Optional whether I say useOptional true or not.  But i probably want to use them anyway....

### PR DESCRIPTION
**Pull request by slarti-b**
### PR checklist

- [ X] Read the [contribution guidelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [ X] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [ X] Filed the PR against the correct branch: `3.0.0` branch for changes related to OpenAPI spec 3.0. Default: `master`.
- [X ] Copied the [technical committee](https://github.com/swagger-api/swagger-codegen/#swagger-codegen-technical-committee) to review the pull request if your PR is targeting a particular programming language. @bbdouglas @JFCote @sreeshas @jfiala @lukoyanov @cbornet 

### Description of the PR

Fixed data type for optional parameters in delegate

When generating using the delegatePattern and including an optional parameter, the API class used a type of Optional<T> while the delegate used just T.  This resulted in a Java error in the generated code when trying to call one from the other.

Fixed by using correct data type in delegate

Fix #7179

### CODE REVIEW

File: samples/server/petstore/spring-mvc-j8-async/src/main/java/io/swagger/api/AnotherFakeApi.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /**
- * NOTE: This class is auto generated by the swagger code generator program (2.3.0-SNAPSHOT).
+ * NOTE: This class is auto generated by the swagger code generator program (unset).
```
Comment (cbornet) [inline samples/server/petstore/spring-mvc-j8-async/src/main/java/io/swagger/api/AnotherFakeApi.java]: Weird...

Comment (slarti-b) [reply samples/server/petstore/spring-mvc-j8-async/src/main/java/io/swagger/api/AnotherFakeApi.java]: Hmm... there are somewhat more update scripts for linux than windows.  Are the windows ones maintained?  I'll try the linux ones...

Comment (cbornet) [reply samples/server/petstore/spring-mvc-j8-async/src/main/java/io/swagger/api/AnotherFakeApi.java]: Yes, the windows scripts are broken

**Comment by slarti-b** [comment]
I tried running the spring-all-petstore.sh script (I run on windows, but with a cygwin shell for my sanity).  It still seems to be setting the version to "unset".  Also, if I then `mvn clean package install` then it fails with

> Failed to execute goal org.eclipse.jetty:jetty-maven-plugin:9.2.9.v20150224:start (start-jetty) on project swagger-generator: Failure: Address already in use: bind

I have therefore reverted the sample updates done by the broken windows script, but not pushed the ones done by the linux scripts either, since they don't seem to be right either.

**Comment by slarti-b** [comment]
Do I still need to do anything with the scripts, or can this be considered as it is?

**Comment by slarti-b** [comment]
@cbornet: Do I need to do anything more with the scripts here?  I merged in from master (to keep myself up to date) and the CI tests pass.

**Comment by cbornet** [comment]
LGTM

**Pull request merged**